### Imports & Paths

In [1]:
import pandas as pd
import numpy as np
import json
from pathlib import Path

### Load PolitiFact Dataset

In [2]:
print("Loading PolitiFact Dataset")

politifact_path = Path('/Users/satwik/Documents/GitHub/UF-EEE6778-Fall25-TermProject/data/raw/politifact_factcheck_data.json')

politifact_data = []
try:
    with open(politifact_path, "r") as f:
        politifact_data = json.load(f)
except json.JSONDecodeError:
    print("Loading as JSONL format (one JSON object per line)...")
    with open(politifact_path, "r") as f:
        for line in f:
            line = line.strip()
            if line:
                politifact_data.append(json.loads(line))

df_politifact = pd.DataFrame(politifact_data)
print(f"PolitiFact Dataset loaded: {df_politifact.shape}")
df_politifact.head(20)

Loading PolitiFact Dataset
Loading as JSONL format (one JSON object per line)...
PolitiFact Dataset loaded: (21152, 8)


,verdict,statement_originator,statement,statement_date,statement_source,factchecker,factcheck_date,factcheck_analysis_link
0,true,Barack Obama,John McCain opposed bankruptcy protections for...,6/11/2008,speech,Adriel Bettelheim,6/16/2008,https://www.politifact.com/factchecks/2008/jun...
1,false,Matt Gaetz,"""Bennie Thompson actively cheer-led riots in t...",6/7/2022,television,Yacob Reyes,6/13/2022,https://www.politifact.com/factchecks/2022/jun...
2,mostly-true,Kelly Ayotte,"Says Maggie Hassan was ""out of state on 30 day...",5/18/2016,news,Clay Wirestone,5/27/2016,https://www.politifact.com/factchecks/2016/may...
3,false,Bloggers,"""BUSTED: CDC Inflated COVID Numbers, Accused o...",2/1/2021,blog,Madison Czopek,2/5/2021,https://www.politifact.com/factchecks/2021/feb...
4,half-true,Bobby Jindal,"""I'm the only (Republican) candidate that has ...",8/30/2015,television,Linda Qiu,8/30/2015,https://www.politifact.com/factchecks/2015/aug...
5,true,S.E. Cupp,"""There are actually only 30 countries that pra...",8/23/2015,news,Will Cabaniss,8/23/2015,https://www.politifact.com/factchecks/2015/aug...
6,false,Michele Bachmann,"""My husband and I have never gotten a penny of...",6/26/2011,news,Robert Farley,7/1/2011,https://www.politifact.com/factchecks/2011/jul...
7,true,George Turner,"""If you go strictly by the numbers, crime is d...",4/4/2011,news,Willoughby Mariano,4/18/2011,https://www.politifact.com/factchecks/2011/apr...
8,mostly-true,Kent Conrad,"""The American people say, don't touch Social S...",1/23/2011,other,Angie Drobnic Holan,1/27/2011,https://www.politifact.com/factchecks/2011/jan...
9,true,Mandela Barnes,"""Since 1978, CEO compensation rose over 1,000%...",5/23/2021,social_media,Madeline Heim,6/2/2021,https://www.politifact.com/factchecks/2021/jun...


### Load Snopes Dataset

In [3]:
print("Loading Snopes Dataset")

snopes_path = Path('/Users/satwik/Documents/GitHub/UF-EEE6778-Fall25-TermProject/data/raw/snopeswithsum.csv')
df_snopes = pd.read_csv(snopes_path)

print(f"Snopes Dataset loaded: {df_snopes.shape}")
df_snopes.head(20)

Loading Snopes Dataset
Snopes Dataset loaded: (4550, 9)


,question,comment,claim,rate,what's true,what's false,what's unknown,origin,summary
0,Did Kamala Harris Support Abortion Until the T...,The Democratic vice-presidential candidate's p...,"U.S. Sen. Kamala Harris, D-Calif., supports th...",Mixture,"U.S. Sen. Kamala Harris, D-Calif, has maintain...",She has not explicitly stated that she support...,NaN,Debates surrounding abortion frequently resurf...,"U.S. Sen. Kamala Harris, D-Calif., has been to..."
1,Did Hitler Invent the Inflatable Sex Doll?,Nothing attracts curiosity like a lurid combin...,Adolf Hitler was behind the invention of the f...,False,NaN,NaN,NaN,"In late July 2020, readers shared a 2016 blog ...",Readers shared a 2016 blog post with the headl...
2,Does Texting ‘RBG’ to 50409 Send a Letter to Y...,The automated political advocacy service Resis...,"Texting ""RBG"" to 50409 engages a service calle...",True,NaN,NaN,NaN,"In September 2020, following the death of U.S....",Readers asked Snopes to examine the accuracy a...
3,Pride Parade Fire Hydrant Mishap,An image depicts a man stabbed at an altercati...,An image depicts an overenthusiastic Pride att...,Miscaptioned,NaN,NaN,NaN,Yearly Pride festivities were marked on 26 Jun...,A Twitter user tweeted a photograph of a blood...
4,Swedish Police Hand Out ‘Don’t Touch Me’ Brace...,Police in Sweden have introduced sexual assaul...,Swedish police have introduced anti-sexual ass...,Mixture,Police in Sweden are distributing colored brac...,Swedish police did not create or distribute th...,NaN,"On 30 June 2016, the Federalist Papers web sit...",Police in Sweden have come up with a sure-fire...
5,Did Trump Say He Might Sign an Exec Order Barr...,"""Was he joking?"" has become a frequent questio...","In September 2020, U.S. President Donald Trump...",Mixture,"On Sept. 19, 2020, Trump said at a rally in No...",NaN,It's unclear whether Trump was joking.,"In September 2020, readers asked Snopes to exa...","In September 2020, readers asked Snopes to exa..."
6,Did the White House Watch the Benghazi Attack ...,Claims that top Obama administration officials...,Top Obama administration officials watched the...,False,NaN,NaN,NaN,One of the more common fault-finding rumors as...,Forbes published an opinion piece claiming tha...
7,Do Color Codes on Toothpaste Tubes Identify Th...,The colored squares on toothpaste tubes suppos...,The colored squares on toothpaste tubes identi...,False,NaN,NaN,NaN,A popular item about markings on toothpaste tu...,The colored squares or rectangular bars found ...
8,Did the Last Surviving WWII Veteran March Alon...,"A photograph of the ""last surviving veteran"" o...",A photograph depicts last surviving veteran of...,Miscaptioned,The photograph depicts an elderly gentleman cr...,The man wasn't identified as the last survivin...,NaN,A number of social media pages devoted to attr...,A photo of a crying older gentleman marching a...
9,Did Ruth Bader Ginsburg Say that Pedophilia Wa...,A 1974 report co-authored by Ruth Bader Ginsbu...,Ruth Bader Ginsburg said that pedophilia was g...,False,NaN,NaN,NaN,The language in a 1974 report that was co-auth...,Ruth Bader Ginsburg was accused of wanting to ...


### Preprocess and Standardize Columns

In [4]:
# Politifact Preprocessing & Standardization
df_politifact = df_politifact.rename(columns={
    "statement": "claim_text",
    "verdict": "verdict_original",
    "statement_originator": "originator",
    "factcheck_analysis_link": "url"
})
df_politifact["dataset_source"] = "PolitiFact"
df_politifact["summary"] = None
df_politifact["claim_id"] = [f"P{i:05d}" for i in range(1, len(df_politifact)+1)]

mapping_pf = {
    "true": "Likely True",
    "mostly-true": "Likely True",
    "half-true": "Uncertain",
    "mostly-false": "Likely False",
    "false": "Likely False",
    "pants-fire": "Likely False"
}
df_politifact["verdict_mapped"] = df_politifact["verdict_original"].str.lower().map(mapping_pf)

df_politifact = df_politifact[
    ["claim_id", "claim_text", "verdict_original", "verdict_mapped",
     "summary", "originator", "url", "dataset_source"]
]

# Snopes Preprocessing & Standardization
df_snopes = df_snopes.rename(columns={
    "claim": "claim_text",
    "rate": "verdict_original",
    "summary": "summary"
})
df_snopes["originator"] = None
df_snopes["url"] = None
df_snopes["dataset_source"] = "Snopes"
df_snopes["claim_id"] = [f"S{i:05d}" for i in range(1, len(df_snopes)+1)]

mapping_snopes = {
    "true": "Likely True",
    "mostly true": "Likely True",
    "false": "Likely False",
    "mostly false": "Likely False",
    "mixture": "Uncertain",
    "unproven": "Uncertain",
    "miscaptioned": "Uncertain",
    "legend": "Uncertain",
    "outdated": "Uncertain",
    "scam": "Likely False",
    "satire": "Uncertain",
    "research in progress": "Uncertain",
    "correct attribution": "Likely True",
    "misattributed": "Likely False"
}
df_snopes["verdict_mapped"] = df_snopes["verdict_original"].str.lower().map(mapping_snopes)

df_snopes = df_snopes[
    ["claim_id", "claim_text", "verdict_original", "verdict_mapped",
     "summary", "originator", "url", "dataset_source"]
]

### Merge and Save Unified Dataset

In [5]:
# Merge both datasets
df_merged = pd.concat([df_politifact, df_snopes], ignore_index=True)
print(f"Merged dataset shape: {df_merged.shape}")

# Check summary
print("\nNull value summary:")
print(df_merged.isnull().sum())

print("\nVerdict Distribution:")
print(df_merged['verdict_mapped'].value_counts(dropna=False))

# Save
output_path = Path('/Users/satwik/Documents/GitHub/UF-EEE6778-Fall25-TermProject/data/processed/merged_factcheck_dataset.csv')
output_path.parent.mkdir(parents=True, exist_ok=True)
df_merged.to_csv(output_path, index=False)
print(f"\n Unified dataset saved to:\n{output_path}")

Merged dataset shape: (25702, 8)

Null value summary:
claim_id                0
claim_text              0
verdict_original        0
verdict_mapped        162
summary             21152
originator           4550
url                  4550
dataset_source          0
dtype: int64

Verdict Distribution:
verdict_mapped
Likely False    14390
Likely True      6526
Uncertain        4624
NaN               162
Name: count, dtype: int64

 Unified dataset saved to:
/Users/satwik/Documents/GitHub/UF-EEE6778-Fall25-TermProject/data/processed/merged_factcheck_dataset.csv


In [6]:
#Display of Merged Dataset
df_merged.head(50)

,claim_id,claim_text,verdict_original,verdict_mapped,summary,originator,url,dataset_source
0,P00001,John McCain opposed bankruptcy protections for...,true,Likely True,None,Barack Obama,https://www.politifact.com/factchecks/2008/jun...,PolitiFact
1,P00002,"""Bennie Thompson actively cheer-led riots in t...",false,Likely False,None,Matt Gaetz,https://www.politifact.com/factchecks/2022/jun...,PolitiFact
2,P00003,"Says Maggie Hassan was ""out of state on 30 day...",mostly-true,Likely True,None,Kelly Ayotte,https://www.politifact.com/factchecks/2016/may...,PolitiFact
3,P00004,"""BUSTED: CDC Inflated COVID Numbers, Accused o...",false,Likely False,None,Bloggers,https://www.politifact.com/factchecks/2021/feb...,PolitiFact
4,P00005,"""I'm the only (Republican) candidate that has ...",half-true,Uncertain,None,Bobby Jindal,https://www.politifact.com/factchecks/2015/aug...,PolitiFact
5,P00006,"""There are actually only 30 countries that pra...",true,Likely True,None,S.E. Cupp,https://www.politifact.com/factchecks/2015/aug...,PolitiFact
6,P00007,"""My husband and I have never gotten a penny of...",false,Likely False,None,Michele Bachmann,https://www.politifact.com/factchecks/2011/jul...,PolitiFact
7,P00008,"""If you go strictly by the numbers, crime is d...",true,Likely True,None,George Turner,https://www.politifact.com/factchecks/2011/apr...,PolitiFact
8,P00009,"""The American people say, don't touch Social S...",mostly-true,Likely True,None,Kent Conrad,https://www.politifact.com/factchecks/2011/jan...,PolitiFact
9,P00010,"""Since 1978, CEO compensation rose over 1,000%...",true,Likely True,None,Mandela Barnes,https://www.politifact.com/factchecks/2021/jun...,PolitiFact


Since we have 162 Null rows in 'verdict_mapped' column which is a Critical column in both Model Training & Retrieval we will drop those rows as it accounts only 0.6% of the total data.

In [7]:
# Drop rows where verdict_mapped is NaN
df_merged = df_merged.dropna(subset=['verdict_mapped']).reset_index(drop=True)

# Confirm cleanup
print(f"After dropping NaN verdicts: {df_merged.shape[0]} rows remain")
print(df_merged['verdict_mapped'].value_counts(dropna=False))

After dropping NaN verdicts: 25540 rows remain
verdict_mapped
Likely False    14390
Likely True      6526
Uncertain        4624
Name: count, dtype: int64


In [8]:
# Check summary (For Verification Purpose)
print("\nNull value summary:")
print(df_merged.isnull().sum())


Null value summary:
claim_id                0
claim_text              0
verdict_original        0
verdict_mapped          0
summary             21152
originator           4388
url                  4388
dataset_source          0
dtype: int64


Since we do not have any Null values in Columns critical for Training 'claim_text', 'verdict_original' & 'verdict_mapped' we can proceed with the next steps.

In [9]:
df_merged.head(25)

,claim_id,claim_text,verdict_original,verdict_mapped,summary,originator,url,dataset_source
0,P00001,John McCain opposed bankruptcy protections for...,true,Likely True,None,Barack Obama,https://www.politifact.com/factchecks/2008/jun...,PolitiFact
1,P00002,"""Bennie Thompson actively cheer-led riots in t...",false,Likely False,None,Matt Gaetz,https://www.politifact.com/factchecks/2022/jun...,PolitiFact
2,P00003,"Says Maggie Hassan was ""out of state on 30 day...",mostly-true,Likely True,None,Kelly Ayotte,https://www.politifact.com/factchecks/2016/may...,PolitiFact
3,P00004,"""BUSTED: CDC Inflated COVID Numbers, Accused o...",false,Likely False,None,Bloggers,https://www.politifact.com/factchecks/2021/feb...,PolitiFact
4,P00005,"""I'm the only (Republican) candidate that has ...",half-true,Uncertain,None,Bobby Jindal,https://www.politifact.com/factchecks/2015/aug...,PolitiFact
5,P00006,"""There are actually only 30 countries that pra...",true,Likely True,None,S.E. Cupp,https://www.politifact.com/factchecks/2015/aug...,PolitiFact
6,P00007,"""My husband and I have never gotten a penny of...",false,Likely False,None,Michele Bachmann,https://www.politifact.com/factchecks/2011/jul...,PolitiFact
7,P00008,"""If you go strictly by the numbers, crime is d...",true,Likely True,None,George Turner,https://www.politifact.com/factchecks/2011/apr...,PolitiFact
8,P00009,"""The American people say, don't touch Social S...",mostly-true,Likely True,None,Kent Conrad,https://www.politifact.com/factchecks/2011/jan...,PolitiFact
9,P00010,"""Since 1978, CEO compensation rose over 1,000%...",true,Likely True,None,Mandela Barnes,https://www.politifact.com/factchecks/2021/jun...,PolitiFact
